### DISCUSSION TITLE
Close all pipelines in cluster using CLI

### DISCUSSION THREAD

**Discussion by Montura** [apache/ozone#6556] [Ideas]
For now ozone CLI allows to [get pipelines list](https://github.com/apache/ozone/blob/master/hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/pipeline/ListPipelinesSubcommand.java#L49) or [close pipeline](https://github.com/apache/ozone/blob/master/hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/pipeline/ClosePipelineSubcommand.java) by `PipelineId`. 

I'd like to add a new command to close all pipelines (or some filtered pipeline subset):
1. get pipeline list from `ScmClient`
2. filter `CLOSED` pipelines (to close only `OPEN` and `ALLOCATED`)
3. filter pipelines by `REPLICATION` or `REPLICATION_TYPE` or `REPLICATION_FACTOR` (as in [ListPipelinesSubcommand](https://github.com/apache/ozone/blob/master/hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/pipeline/ListPipelinesSubcommand.java#L103) done) 
4. close the remain pipelines set

UPD: added [PR](https://github.com/apache/ozone/pull/6577)

**Comment by swamirishi** [comment]
@Montura Can you file a jira for this.

**Reply by Montura** [reply]
Sure, I'll provide Jira and PR, if my suggested solution is ok for contributors

**Comment by jojochuang** [comment]
By the way, you can do that using the following script:

```shell
ozone admin pipeline list | grep OPEN | cut -d ',' -f1 | cut -d ":" -f2 | cut -d " " -f2  | while read line; do echo $line | ozone admin pipeline close $line ; done
```

**Reply by Montura** [reply]
Thanks, but I suppose `ozone admin pipeline closeAll` will be more user-friendly

**Reply by smengcl** [reply]
or `ozone admin pipeline close --all` :)

**Reply by Montura** [reply]
@smengcl, better solution then adding a new subcommand I suppose

**Comment by errose28** [comment]
It seems like it would be a nice addition, but can you share more details about the use case? If mass closing of pipelines is so frequent that it needs a convenient command, there is probably something else wrong with the cluster.

**Reply by Montura** [reply]
According to discussions in https://github.com/apache/ozone/discussions/6248 for now ozone doesn't have a mechanism for `fair pipeline distribution`. And in some cases it's important to have `fairly distributed pipelines between the datanodes`.

So, I found [possible solution](https://github.com/apache/ozone/discussions/6248#discussioncomment-8557298) for it by specifying `the minimun number of datanodes required to exit SCM safe mode` and then create pipelines. But this approach does not work very well in some fault tolerance scenarios.

So, the obvious way to achieve the `desired pipeline distribution` is to **close all pipelines that were created during cluster startup**. At this point **all datanodes are (hopefully) healthy**. Then `BackgroundPipelineCreator` can create pipelines using all healthy datanodes instead of creating as much as possible pipelines on **the first 3 started datanodes**.

**Comment by ivandika3** [comment]
Closed since https://github.com/apache/ozone/pull/6577 has been merged.
